### ISSUE TITLE
Start-Process: allow redirecting std output/error to console stream

### ISSUE THREAD

**Issue by aetonsi** [PowerShell/PowerShell#18026]
### Summary of the new feature / enhancement

Hello.
I am trying to start a process in a new window with `Start-Process` and i'm `-wait`ing for its execution.

Then i need to process its output. I would like to do it all in memory, without writing to disk, **as the output is confidential**. Unfortunately at the moment this doesn't seem to be possible (correct me if i'm wrong) without resorting to .NET implementations like [this](https://stackoverflow.com/a/8762068)...

Currently, this is the best you can do:

```powershell
$output = & {start-process powershell "read-host 'Enter string: '" -Wait -RedirectStandardOutput output.txt ; get-content output.txt}
```
In this example,
- `powershell "read-host 'Enter string: '"` is my command with confidential output.
- `-Wait` waits for my command to exit
- `-RedirectStandardOutput` permits the redirection to a file
- `get-content` simply reads back that file.

### Proposed technical implementation details (optional)

A possible solution would be a parameter like `-RedirectStandardOutputToStream` / `-RedirectStandardErrorToStream`

```powershell
$output = & {start-process powershell "read-host 'Enter string: '" -wait -RedirectStandardOutputToStream 1}
```

Where `1` is the ID of the [success stream](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_redirection?view=powershell-7.2#redirectable-output-streams).

---

Or, more simply:

```powershell
$output = & {start-process powershell "read-host 'Enter string: '" -wait -GetStandardOutput}
```
... with the corresponding `-GetStandardError`

**Comment by aetonsi** [comment]
Here i posted a workaround leveraging a temp file: [aetonsi/pwsh__StartProcessWaitRedirectToStream](https://github.com/aetonsi/pwsh__StartProcessWaitRedirectToStream/commit/d4c4b51c50d9c2840199ba800913608c68c62d61)

**Comment by jhoneill** [comment]
What are you using start-process at all ? 
 ```$output = powershell "read-host 'I should not ask for parameters to be typed' " ```   or  
```$output = powershell "read host 'hyphen missing will cause error' " 2> null  ```   if you want to remove error output.  

If it is a command line program start-process will (by default) run in a new window, not receive output and not wait for the program it has started to finish. You can specify Wait or NoNewWindow but you still need to work around the output not coming back 
 
```$output = Invoke-Command -ScriptBlock {powershell "read-host 'I should not ask for parameters to be typed' " }``` 
Also works the way you want, I think.

**Comment by aetonsi** [comment]
I need my command to be run in a separate window, that's why i resorted to `start-process`. Is there some other way?

**Comment by jborean93** [comment]
Not without using dotnet but that’s the beauty of PowerShell, you can wrap all that in your own function to achieve what you want.

**Comment by mklement0** [comment]
Note that there's a previous discussion about generally allowing targeting _variables_ in context where _file paths_ are expected, so that `Start-Process ... -RedirectStandardOutput variable:out` would capture stdout in variable `$out`.

This is not only more flexible than `-RedirectStandardOutput 1`, it also prevents the output from instantly mixing with the caller's streams, which in _asynchronous_ `Start-Process` calls (which are typical) you wouldn't want.

See:
* https://github.com/PowerShell/PowerShell/issues/4332 (original discussion, in the context of `>`)
* https://github.com/PowerShell/PowerShell/issues/5184 (later discussion, in the context of `Start-Process`)

**Comment by aetonsi** [comment]
@mklement0 Mmh i see, that would be cool. Not the same as what i was picturing (`-RedirectStandardOutputToStream 1` would print to the success stream, and you _could_ capture it in a variable)... but ultimately that's what i would like to end up to do: save the output to a variable to reuse it.

**Comment by SteveL-MSFT** [comment]
@JamesWTruher I thought you had a proof-of-concept for supporting redirection to the variable provider?

**Comment by JamesWTruher** [comment]
@SteveL-MSFT Indeed I do: https://github.com/PowerShell/PowerShell/pull/16497

**Comment by jborean93** [comment]
That doesn’t really help here for Start-Process if I understand it correctly? You can already capture output with just invoking the exe today. Supporting Start-Process redirection to a var would be quite tricky unless it meant -Wait was always implicit as it would require a background task to be running which could mutate the var as it’s being read.

**Comment by jhoneill** [comment]
> That doesn’t really help here for Start-Process if I understand it correctly?

I think in this case there is output from Start-Process itself that can be redirected. It would need start-process to allow 
`-RedirectStandardOutput variable:foo`  but that reports "Cannot open file because the current provider (Microsoft.PowerShell.Core\Variable) cannot open a file."

**Comment by aetonsi** [comment]
>  You can already capture output with just invoking the exe today

Yes but AFAIK you cannot start a process _in a new window_, _wait_ for it and capture the output

**Comment by jborean93** [comment]
Why would you want to, that would mean you have a blank window with no output. When you redirect a std pipe in a process you cause that output to no longer appear. You would have to somehow split the output to go to a pipe you can capture but also have it output to the console at the same time which isn’t really easy to do at all.

**Comment by mklement0** [comment]
@jborean93, it's a good point that redirecting to _variables_ could only work meaningfully with (possibly implied) `-Wait`

> You can already capture output with just invoking the exe today

The original issue, #4332, was focused on `2>` redirections to allow you to capture _stderr_ output in a variable, as the native analog to `-ErrorVariable` - direct invocation of an external program currently doesn't support that.

In the context of (synchronous use of) `Start-Process`:

> Why would you want to [....]?

They're certainly not common uses cases, but two come to mind (in both cases, `-WindowStyle Hidden` can be used to hide the new window):

* If you want to run a process _as another user_, with `-Credential` (on Windows), which cannot be done without `Start-Process` and invariably involves a new window.

* If you need to "sandbox" invocation of a PowerShell script that may interfere with the state of the calling session if invoked directly. (Arguably, though, `Start-Job` could be used instead and is the better choice.)

**Comment by jhoneill** [comment]
> Why would you want to, that would mean you have a blank window with no output. When you redirect a std pipe in a process you cause that output to no longer appear.

@aetonsi  I didn't think of this when you said to my suggestion before that you wanted it its own window. Why do do you want a **blank** window?

**Comment by mklement0** [comment]
@jhoneill:

* The uses cases in my previous comment imply _hidden_ execution, in which case the question of a blank window is moot.

* Even if the launched window runs visibly, the window is only _invariably_ blank if you redirect _both_ stdout and stderr.

The bottom line is: 

Are there legitimate use cases for _synchronous_ use of `Start-Process` where capturing stdout _and/or_ stderr output in-memory is useful?

**Comment by jborean93** [comment]
> The original issue, https://github.com/PowerShell/PowerShell/issues/4332, was focused on 2> redirections to allow you to capture stderr output in a variable, as the native analog to -ErrorVariable - direct invocation of an external program currently doesn't support that.

It’s not pretty but you can do this today

```powershell
$stdout = $null
$stderr = . { my.exe args | Set-Variable stdout } 2>&1 | ForEach-Object ToString
```

The ForEach-Object ToString` is only needed for WinPS because on the older versions these objects are error records and you want the string value instead.

> If you want to run a process as another user, with -Credential (on Windows), which cannot be done without Start-Process and invariably involves a new window.

I will concede this is a valid point, thank you I didn’t think off this.

> If you need to "sandbox" invocation of a PowerShell script that may interfere with the state of the calling session if invoked directly. (Arguably, though, Start-Job could be used instead and is the better choice.)

I’m not sure why this needs Start-Process. You can certainly just do `pwsh { … }` or as you said `Start-Job` to do this in such a nicer way. PowerShell will even deal with complex objects with both of these cases.

Still I feel like the better option is to add NoteProperties to the return object if `-PassThru -Wait` is set and redirection needed. The UX here is pretty poor but it ensures that there’s no concurrency problem, i.e. the output is only accessible once nothing else will add to it.

**Comment by mklement0** [comment]
@jborean93 

>  I’m not sure why this needs `Start-Process`. You can certainly just do `pwsh { … }`

Good point: `pwsh { … }` is definitely the way to go here, which again calls for a simple `2>` solution (see below).

> It’s not pretty 

That's precisely the point: _let's make it pretty (easy)_:

```
$stdout = my.exe args 2>variable:stderr
```

As for the least ugly version right now:

```
# Capture stderr output in $stderr
$stdout, $stderr = (my.exe args).Where({ $_ -is [string] }, 'Split')
```

But all of that relates to _direct_ invocation, which I think is the primary use case for the proposed enhancement.

I don't feel strongly about the `Start-Process` case.

I'm unclear on what you mean by adding NoteProperties.

**Comment by aetonsi** [comment]
> Why would you want to, that would mean you have a blank window with no output. When you redirect a std pipe in a process you cause that output to no longer appear. You would have to somehow split the output to go to a pipe you can capture but also have it output to the console at the same time which isn’t really easy to do at all.

@mklement0 @jhoneill 
Afaik (i'm new to PS so i might be wrong), the window would be blank only if all of the streams with any output are redirected.
But for example i have a simple program that writes to streams # 1, 2, 3 and 6 (`write-host`) and i need just one of those streams redirected to a variable. ATM i can do it by using a support file via `-RedirectStandardOutput` or `-RedirectStandardError`...
Again, if you're wondering why i _need_ `start-process`, it is simply to open a new window. Why? for example because the original process has no window (because it has been started by another c# program with `.StartInfo.WindowStyle = ProcessWindowStyle.Hidden`)

And now that i think about it, this can work only for streams 1/2, number 3-6 are unmanageable. Even with a syntactic sugar like `$stdout = my.exe args 6>variable:stdinfo`, you couldnt apply this to `start-process`.

**Comment by mklement0** [comment]
@aetonsi, yes, when you call a script via the PowerShell CLI, the stdout and stderr streams are the only ones available, as with any child process. By default, PowerShell only sends its error stream to stderr (curiously, only when stderr is _redirected_), all others (success output, warnings, verbose messages, ...) are output to stdout (which is problematic - see #7989)

The only way to distinguish between PowerShell's [output streams](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Output_Streams) via a [CLI](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Pwsh) call would be to use `-OutputFormat xml` in order to receive CLIXML data _via stdout_ (CLIXML is an XML-based object-serialization format). Deserializing this data with `[System.Management.Automation.PSSerializer]::Deserialize()`, would allow you to separate the deserialized objects into their streams of origin by their type.

In other words: this approach indeed would leave your window invariably blank, because _all_ output would be communicated via stdout.

**Comment by mmis1000** [comment]
If someone still read this, it seems to be possible to workaround this by read stream line by line by invoke underlying .net class directly. Although you can only get either stdout or stderr

```powershell
$pinfo = New-Object System.Diagnostics.ProcessStartInfo
$pinfo.FileName = "ping"
# can't read both
# $pinfo.RedirectStandardError = $true
$pinfo.RedirectStandardOutput = $true
$pinfo.UseShellExecute = $false
$pinfo.Arguments = "8.8.8.8"
$p = New-Object System.Diagnostics.Process
$p.StartInfo = $pinfo

$p.Start() | Out-Null

while (($line = $p.StandardOutput.ReadLine()) -ne $null) {
    if ($line -ne '') {
        Write-Host (Get-Date -Format "yyyy-MM-dd HH:mm:ss") 'line: ' $line
    } else {
        Write-Host (Get-Date -Format "yyyy-MM-dd HH:mm:ss") 'line: <blank>'
    }
}

$p.WaitForExit()

Write-Host "exit code: " $p.ExitCode
```

**Comment by aetonsi** [comment]
Mhh so still nothing for streams 3 to 6....
